.rubocop.yml

```diff
@@ -56,3 +61,44 @@ Style/AndOr:
   # or completely (always).
   # They read better, more like normal English.
   Enabled: false
+
+Layout/EmptyLinesAroundAttributeAccessor:
+  Enabled: true
+
+Layout/SpaceAroundMethodCallOperator:
+  Enabled: true
+
+Lint/RaiseException:
+  Enabled: true
+
+Lint/StructNewOverride:
+  Enabled: true
+
+Style/ExponentialNotation:
+  Enabled: true
+
+Style/HashEachMethods:
+  Enabled: true
+
+Style/HashTransformKeys:
+  Enabled: true
+
+Style/HashTransformValues:
+  Enabled: true
+
+Style/SlicingWithRange:
+  Enabled: true
+
+# Until we drop support for Rails 4.2, this can be different Rails 4.2 can work
+# with ruby 2.2 and we don't know of any obvious problems with older versions of
+# ruby, we just don't support them. Making the minimum ruby version match the
+# oldest supported ruby will prevent people from installing the gem in older
+# applications, which we don't want to do.
```

Comment:
Making sure I understand the problem - it looks like this version of Rubocop wants `TargetRubyVersion` >= 2.4. From my fiddling:

```
› rubocop
Error: RuboCop found unsupported Ruby version 2.2 in `TargetRubyVersion`
 parameter (in .rubocop.yml). 2.2-compatible analysis was dropped after
version 0.69.
Supported versions: 2.4, 2.5, 2.6, 2.7
```

Since Rails 4.2 supports Ruby >= 2.2, we want to have the `gemspec` show support for `2.2` and adding this exclusion helps prevent Rubocop from complaining about that.

I'm not 100% sure I understood the comment, but tried to piece it together based on my testing and make suggestions:
```suggestion
# Until we drop support for Rails 4.2, this can be different. Rails 4.2 can
# work with Ruby 2.2 and we don't know of any obvious problems with older
# versions of Ruby (we just don't support them). Making the minimum Ruby
# version match the oldest supported version Ruby from Rubocop would require
# specifying a newer version of Ruby in the `gemspec`. This would prevent
# people from installing the gem in older applications, which we don't want to
# do.
```

I may be completely misunderstanding it though.

Reply:
Also, the gemspec seems to indicate 2.3 is the minimum version:

https://github.com/devise-security/devise-security/blob/63b8b2ca9cbff71d75f329337f04cc105a674e83/devise-security.gemspec#L27

If the goal is >= 2.2, I think the changes to use the safe navigation operator will need to be reverted since it looks like that was [added in Ruby 2.3](https://aaronlasseigne.com/2016/01/04/rubys-new-safe-navigation-not-equal-operator/)

Reply:
We dropped support for Ruby 2.2 in bf60b014d935c47aaece86cb5dd341fd8f120a37, which was quite a while ago. So I'm fine with the Ruby 2.3 language features. The gemspec will prevent people using Ruby 2.2 from installing it.  Rails 4.2 can run on Ruby 2.3+ so that's fine. I just don't want to bump the lower limit on the Ruby version at this point because anyone using Rails 4.2 and Ruby 2.4 (for example) would not be able to install the gem. When we officially drop support for Rails 4.2 in v0.16.0 I will consider bumping this.  As a rule, I would rather not bump the gemspec version requirement unless we know we are using language features that need it.

Reply:
Thank you for the explanation. I think that confirms my understanding of the comment and what the `Gemspec/RequiredRubyVersion` cop is trying to do.